nautobot/core/api/serializers.py

```diff
@@ -315,3 +399,195 @@ class BulkOperationIntegerIDSerializer(serializers.Serializer):
+                    existing_objects_2 = [assoc.get_peer(instance) for assoc in existing_associations_2]
+                    existing_associations = list(existing_associations_1) + list(existing_associations_2)
+                    existing_objects = existing_objects_1 + existing_objects_2
+
+                add_objects = []
+                remove_assocs = []
+
+                for obj, assoc in zip(existing_objects, existing_associations):
+                    if obj not in expected_objects:
+                        remove_assocs.append(assoc)
+                for obj in expected_objects:
+                    if obj not in existing_objects:
+                        add_objects.append(obj)
+
+                for add_object in add_objects:
+                    if "request" in self.context and not self.context["request"].user.has_perm(
+                        "extras.add_relationshipassociation"
+                    ):
+                        raise PermissionDenied("This user does not have permission to create RelationshipAssociations.")
+                    if other_side != RelationshipSideChoices.SIDE_SOURCE:
+                        assoc = RelationshipAssociation(
+                            relationship=relationship,
+                            source_type=relationship.source_type,
+                            source_id=instance.id,
+                            destination_type=relationship.destination_type,
+                            destination_id=add_object.id,
+                        )
+                    else:
+                        assoc = RelationshipAssociation(
+                            relationship=relationship,
+                            source_type=relationship.source_type,
+                            source_id=add_object.id,
+                            destination_type=relationship.destination_type,
+                            destination_id=instance.id,
+                        )
+                    assoc.validated_save()  # enforce relationship filter logic, etc.
+                    logger.debug("Created %s", assoc)
+
+                for remove_assoc in remove_assocs:
+                    if "request" in self.context and not self.context["request"].user.has_perm(
+                        "extras.delete_relationshipassociation"
+                    ):
+                        raise PermissionDenied("This user does not have permission to delete RelationshipAssociations.")
+                    logger.debug("Deleting %s", remove_assoc)
+                    remove_assoc.delete()
+
+    def get_field_names(self, declared_fields, info):
+        """Ensure that "relationships" is always included as an opt-in field."""
+        fields = list(super().get_field_names(declared_fields, info))
+        self.extend_field_names(fields, "relationships", opt_in_only=True)
+        return fields
+
+
+class NotesSerializerMixin(BaseModelSerializer):
```

Comment:
nit: move to `nautobot.core.api.mixins`?

Reply:
Same comment above